### PR TITLE
languagedetection: Read binID directly from exe

### PR DESCRIPTION
**Pull request by vitkyrka**
### What does this PR do?

The current code tries to first read the target of the exe symbolic link and then appends that path to the proc's root, and then tries to stat(2) that.  That doesn't work if the binary is not available inside the proc's root for any reason.  However, stat(2)ing the exe itself always works, so do that instead.

For example avahi-daemon works like this:

```
 $ docker run --privileged -it --rm --platform linux/amd64 -v
 /proc/:/host/proc/:ro  --entrypoint /bin/sh alpine
 / # ls -l /host/proc/1640/exe
 lrwxrwxrwx    1 root     root             0 Aug 28 09:09
 /host/proc/1640/exe -> /usr/sbin/avahi-daemon
 / # ls -l /host/proc/1640/root
 lrwxrwxrwx    1 root     root             0 Aug 28 12:38
 /host/proc/1640/root -> /etc/avahi
 / # ls /host/proc/1640/root/usr/sbin/avahi-daemon
 ls: /host/proc/1640/root/usr/sbin/avahi-daemon: No such file or directory
 / # stat -L /host/proc/1640/exe
   File: /host/proc/1640/exe
   Size: 154120          Blocks: 304        IO Block: 4096   regular file
 Device: fc04h/64516d    Inode: 3932513     Links: 1
```

On the host:
```
 $ stat /usr/sbin/avahi-daemon
   File: /usr/sbin/avahi-daemon
   Size: 154120    	Blocks: 304        IO Block: 4096   regular file
 Device: fc04h/64516d	Inode: 3932513     Links: 1
```

As can be seen, stat(2) on the exe gets the correct details of the binary while the current method doesn't work.

### Motivation

Discovered due to debug prints when running service discovery tests :
```
=== RUN   TestInjectedServiceName
    mock.go:33: 2024-08-28 16:51:01 CEST | DEBUG | (pkg/languagedetection/privileged/privileged_detector.go:77 in DetectWithPrivileges) | failed to get binID: stat binary path /proc/1640/root/usr/sbin/avahi-daemon: no such file or directory
```

### CODE REVIEW

File: pkg/languagedetection/privileged/privileged_detector.go [hunk 1]
```diff
@@ -109,17 +109,11 @@ func MockPrivilegedDetectors(t *testing.T, newDetectors []languagemodels.Detecto
...
-	if err != nil {
-		return binaryID{}, fmt.Errorf("readlink %s: %v", exePath, err)
-	}
-
-	binPath = filepath.Join(procPath, "root", binPath)
 
 	var stat syscall.Stat_t
-	err = syscall.Stat(binPath, &stat)
+	err := syscall.Stat(exePath, &stat)
 	if err != nil {
```
Comment (guyarb) [inline pkg/languagedetection/privileged/privileged_detector.go:115]: nit: 
```suggestion
	
	if err := syscall.Stat(exePath, &stat); err != nil {
```

File: pkg/languagedetection/privileged/privileged_detector_test.go [hunk 2]
```diff
@@ -97,6 +99,38 @@ func TestGetBinID(t *testing.T) {
 	assert.Equal(t, binID1, binID2)
 }
 
+func copyForkExecDelete(t *testing.T, timeout int) *exec.Cmd {
```
Comment (guyarb) [inline pkg/languagedetection/privileged/privileged_detector_test.go:102]: can you add documentation?

File: pkg/languagedetection/privileged/privileged_detector_test.go [hunk 3]
```diff
@@ -97,6 +99,38 @@ func TestGetBinID(t *testing.T) {
...
+	sleepBin := filepath.Join(t.TempDir(), "sleep")
+	require.NoError(t, exec.Command("cp", "/bin/sleep", sleepBin).Run())
+
+	cmd := exec.Command(sleepBin, strconv.Itoa(timeout))
+	err := cmd.Start()
+
+	time.Sleep(250 * time.Millisecond)
+	require.NoError(t, err)
```
Comment (guyarb) [inline pkg/languagedetection/privileged/privileged_detector_test.go:110]:
```suggestion
	require.NoError(t, cmd.Start())

	time.Sleep(250 * time.Millisecond)
```

File: pkg/languagedetection/privileged/privileged_detector_test.go [hunk 4]
```diff
@@ -97,6 +99,38 @@ func TestGetBinID(t *testing.T) {
...
+
+	// cmd.Start() waits for the exec to happen so deleting the file should be
+	// safe immediately. Unclear what the sleep above is for; it's copied from
+	// the other tests.
+	os.Remove((sleepBin))
```
Comment (guyarb) [inline pkg/languagedetection/privileged/privileged_detector_test.go:115]:
```suggestion
	os.Remove(sleepBin)
```

**Review by guyarb — APPROVED** [review]

File: pkg/languagedetection/privileged/privileged_detector_test.go (hunk 2 above)
Comment (vitkyrka) [reply pkg/languagedetection/privileged/privileged_detector_test.go:110]: Added.

**Pull request merged**